### PR TITLE
SLE-689 / SLE-691: Core changes for rule descriptions

### PR DESCRIPTION
**Pull request by thahnen**
## Summary

Changes to the rule description inside the specific view and the settings. The fake data is here to stay while testing and no official analyzer is available for dogfooding!

### CODE REVIEW

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java [hunk 1]
```diff
@@ -28,49 +29,101 @@
...
-  private final Label ruleTypeLabel;
-  private final Label ruleSeverityIcon;
-  private final Label ruleSeverityLabel;
+  private final Label label1;
+  private final Label label2;
+  private final Label label3;
+  private final Label label4;
+  private final Label label5;
+  private final Label label6;
+  private final Label label7;
```
Comment (damien-urruty-sonarsource) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java:57]: I think fields named like this are super hard to follow, and it's very easy to mix them up. If the panels have different structures, why not introduce 2 different classes? One `LegacyRuleHeaderPanel` (that will be deleted at some point) and this class that would be used for the new CCT?

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/SonarLintImages.java [hunk 2]
```diff
@@ -133,6 +130,11 @@ public static Image getSeverityImage(IssueSeverity severity) {
   public static Image getTypeImage(RuleType type) {
     return createImage("type/" + type.name().toLowerCase(Locale.ENGLISH) + ".png");
   }
+  
+  @Nullable
+  public static Image getImpactImage(ImpactSeverity impact) {
+    return createImage("impact/" + impact.name() + ".png");
+  }
```
[the remark this answers is not in the document]
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/SonarLintImages.java:137]: No, we have it cached in the `ImageRegistry`: The first time we call `getImpactImage` it is created, and the other times only retrieved.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleDetailsPanel.java [hunk 3]
```diff
@@ -111,32 +117,32 @@ private void updateScrollCompositeMinSize() {
...
+      
+      ruleHeaderPanel.updateRule(ruleInformation);
 
-      updateHtmlDescription(details.getDescription(), details.getLanguage().getLanguageKey());
-      updateParameters(details);
+      updateHtmlDescription(description, ruleInformation.getLanguage().getLanguageKey());
+      
+      if (ruleInformation instanceof EffectiveRuleDetailsDto) {
+        updateParameters((EffectiveRuleDetailsDto) ruleInformation);
+      }
```
[the remark this answers is not in the document]
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleDetailsPanel.java:145]: Sadly no because we have to run `requestLayout()` and `updateScrollCompositeMinSize()` afterward, therefore moving it to the other method would move this out of the shared method as well. Additionally, the try-catch block also applies to this `updateParameters(...)` call and so it would have to be moved to the separate methods as well.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java (hunk 1 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java:57]: I see the point and I have no idea how to name them better. The issue with having two classes is that we would have to dispose of and create a new one from scratch every time we call `RuleDetailsPanel#updateRule(...)` based on the new CCT or the old CCT.
The **Rule Description view** is sadly already quite slow and flickery due to the disposing of the `RuleDescriptionPanel` every time we update it and I favor stable usability over maintainability in this case. Yes such a naming is hard to follow but with the description and it being a minor and simple class I think it is manageable.
WDYT?

Comment (damien-urruty-sonarsource) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java:57]: Another idea instead of adding/removing the view each time would be to just hide the one not in use: you could create the 2 classes, always add them to the tree and `setVisible` only one of them based on the CCT to use. 
The other argument is that the day we remove the legacy CCT, and I know it's not for tomorrow, it will be simpler to remove the legacy class.
I agree it's a tradeoff, I would still prefer maintainability and readability in this case over efficiency

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleDetailsPanel.java [hunk 4]
```diff
@@ -109,40 +99,52 @@ private void updateScrollCompositeMinSize() {
...
+      ruleHeaderPanel.dispose();
+    }
+    
+    var attributeOptional = ruleInformation.getCleanCodeAttribute();
+    var impacts = ruleInformation.getDefaultImpacts();
+    
+    if (attributeOptional.isPresent() && !impacts.isEmpty()) {
+      ruleHeaderPanel = new RuleHeaderPanel(scrolledContent);
+    } else {
+      ruleHeaderPanel = new LegacyRuleHeaderPanel(scrolledContent);
```
Comment (nquinquenel) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleDetailsPanel.java:143]: I'm not familiar with Eclipse code but if this method is purely for the rule panel (not the issue panel), then we only want to show the new CCT. In any case, we will always have the new analyzer in the upcoming release, so we should expect only new CCT for rules.

The only case with old CCT is with SQ <10.2, which means it's only for issues in connected mode

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java (hunk 1 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleHeaderPanel.java]: Resolved but was only possible by disposing of the `AbstractRuleHeaderPanel`. But this did not cause the big deal I was anticipating so no further decrease of performance on loading a rule description.

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleDetailsPanel.java (hunk 4 above)
Comment (thahnen) [reply org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/rule/RuleDetailsPanel.java:143]: No, everything in Eclipse is shared 😄 This is reused in the **Rule Description view** (for the issues) and the **Rules Configuration settings panel** (for the rules).

**Comment by thahnen** [comment]
@damien-urruty-sonarsource the two code smells will be fixed on the parent branch with SLE-695

File: its/pom.xml [hunk 5]
```diff
@@ -159,6 +159,7 @@
           <forkedProcessTimeoutInSeconds>1800</forkedProcessTimeoutInSeconds>
           <redirectTestOutputToFile>true</redirectTestOutputToFile>
           <systemProperties>
+			<target.platform>${target.platform}</target.platform>
```
Comment (damien-urruty-sonarsource) [inline its/pom.xml:162]: The indentation looks strange here

File: its/src/org/sonarlint/eclipse/its/reddeer/views/RuleDescriptionView.java [hunk 6]
```diff
@@ -50,7 +50,7 @@ public Label getRuleSeverity() {
   }
 
   public Label getRuleKey() {
-    return new DefaultLabel(getCTabItem(), 5);
+    return new DefaultLabel(getCTabItem(), 8);
```
Comment (damien-urruty-sonarsource) [inline its/src/org/sonarlint/eclipse/its/reddeer/views/RuleDescriptionView.java:53]: What does this 8 represent? Maybe you could extract it as a constant and give it a meaningful name?

**Review by damien-urruty-sonarsource — APPROVED** [review]
LGTM

File: its/src/org/sonarlint/eclipse/its/reddeer/views/RuleDescriptionView.java (hunk 6 above)
Comment (thahnen) [reply its/src/org/sonarlint/eclipse/its/reddeer/views/RuleDescriptionView.java:53]: Yeah, this is on my agenda to clarify that magic number when I'll be implementing all the ITs / UTs. It is currently switched to 8 because we have 8 labels for the Rule Description view header (where clean code attribute / impacts / rule key live). 5 will be for the IT with the connection to an old SQ with the legacy header shown.
